### ISSUE TITLE
Optimize read_results

### ISSUE THREAD

**Issue by rikardn** [pharmpy/pharmpy#1138]
@pharmpy-dev-123 found that 5% of unit test time is spent in`read_results`. Perhaps this can be optimized.

**Comment by pharmpy-dev-123** [comment]
Most of the time spent in `read_results` comes from `altair`'s internals (`from_dict` functions).

**Comment by pharmpy-dev-123** [comment]
The issue with `altair` seems to be that we are using a generic parse function that does not need to know about the type of `Chart` it parses before hand. It just tries all possible chart subclasses until it finds one that works. Each attempt consists in running some schema validation which is really slow. See here:

https://github.com/altair-viz/altair/blob/bff48d5a235363a74df881c0d167206a9bbb45e1/altair/vegalite/v4/api.py#L2000-L2006

And where we use it:

https://github.com/pharmpy/pharmpy/blob/4681091c64d3e52e039babf012de4b58005d72c1/src/pharmpy/results.py#L67

We can somewhat alleviate the issue by using a more explicit encoding of the Chart class when writing results, but we cannot circumvent it when the results JSON comes from PsN for instance.

**Comment by pharmpy-dev-123** [comment]
Store the class in the JSON.

**Comment by pharmpy-dev-123** [comment]
Here is what `altair` tries in sequence for a `VConcatChart`:

```
<super: <class 'Chart'>, <Chart object>> from_dict
<super: <class 'Chart'>, <Chart object>> error
<class 'altair.vegalite.v4.api.RepeatChart'> from_dict
<class 'altair.vegalite.v4.api.RepeatChart'> error
<class 'altair.vegalite.v4.api.ConcatChart'> from_dict
<class 'altair.vegalite.v4.api.ConcatChart'> error
<class 'altair.vegalite.v4.api.HConcatChart'> from_dict
<class 'altair.vegalite.v4.api.HConcatChart'> error
<class 'altair.vegalite.v4.api.VConcatChart'> from_dict
SUCCESS!
```

**Comment by rikardn** [comment]
Ok, so if altair had stored the class name it could go directly to the correct schema. Can altair be given that information when it does parsing via `from_json` on the subclasses?

**Comment by pharmpy-dev-123** [comment]
Yes. `altair` does not do it, I guess it prefers to rely on content validation. But we can store which class it was explicitly and then call `class_.from_dict(..., validate=False)` at our own risk.

**Comment by pharmpy-dev-123** [comment]
28.53% after #1347.

**Comment by pharmpy-dev-123** [comment]
31% at d9d04aec

**Comment by pharmpy-dev-123** [comment]
I think this one is a desperate case and maybe something like the following could help: https://docs.pytest.org/en/6.2.x/example/simple.html?highlight=runslow#control-skipping-of-tests-according-to-command-line-option.

Although I would use a `--runfast` or similar approach instead so that CI does not need to have this flag.

**Comment by pharmpy-dev-123** [comment]
That and/or reduce the number of tests that call this function.

**Comment by pharmpy-dev-123** [comment]
This is also related: https://github.com/pytest-dev/pytest-xdist/issues/657

**Comment by rikardn** [comment]
I misunderstood. This is about results.json
